### ISSUE TITLE
Allow disabling inclusion of cache digest fingerprints in asset file names

### ISSUE THREAD

**Issue by rgoldberg** [bertramdev/asset-pipeline#259]
Currently, cache digest fingerprints are always inserted into asset file names.

Please allow disabling the inclusion of cache digest fingerprints in asset file names, by supporting a `Config.groovy` setting named `grails.assets.enableDigests`.

I think that this should be able to be achieved by two small changes:

In the Grails Asset Pipeline plugin, in `_AssetCompile.groovy`, right after:

```
assetConfig.skipNonDigests = true
```

insert:

```
if (config.grails.assets.containsKey('enableDigests')) {
	assetConfig.enableDigests = config.grails.assets.enableDigests
}
```

or, you could insert the following, instead:

```
assetConfig.enableDigests = config.grails.assets.containsKey('enableDigests') ? config.grails.assets.enableDigests : true
```

In asset-pipeline-core (I know it's a separate GitHub project, but I wanted to keep this all in one issue), in `AssetCompiler#createCompressedFiles(..)`, move the following line:

```
File zipFileDigest = new File("${digestedFile.getAbsolutePath()}.gz")
```

to be directly after:

```
if(options.enableDigests as Boolean) {
```

**Comment by davydotcom** [comment]
Can you by chance provide an explanation as to why you would want to do disable digest names in grails as this will break the servlet renderer?

**Comment by rgoldberg** [comment]
I am migrating an existing Grails app from Grails 2.2.4 to 2.5.0, and from Resources to Asset Pipeline.

There are some existing static HTML, JS, & CSS asset files that link to other such files.

These static assets are served by S3 (later will be served via CloudFront), not by the Grails server

While Asset Pipeline will rewrite URLs in CSS files at build time, including injecting cache digest fingerprints into the file names, I was unable to find out how to rewrite URLs in HTML & JS at build time.

It seems like, during runtime in a Grails server, Asset Pipeline will reroute requests for an asset sans cache digest fingerprint to the newest version of that asset, but this won't work in my Grails app since the static assets on S3 are linking to other static assets on S3, so, at runtime, the Grails server isn't involved with serving those assets (Grails just responds to REST calls on Grails Controllers).

If there is some way to rewrite static URLs in static assets to include cache digest fingerprints, then I won't need to remove them from the file names.

Please see the following StackOverflow question with no answers:

http://stackoverflow.com/questions/30878976/grails-asset-pipeline-urls-with-cache-digest-in-file-name-within-static-assets

**Comment by rgoldberg** [comment]
I did not intend to break the servlet renderer, so that would have to be updated to work with the proposed changes.

I couldn't find a Servlet in either of the repos.  I did find `AssetPipelineFilter`, however, in both repos under different packages.  Is that what you're referring to?

**Comment by davydotcom** [comment]
We do no such thing as redirecting requests sans digest. What we do is find the comparable file and assign a proper ETag header but there is no "redirect" concept for sans digested assets, you simply will get the asset file

**Comment by davydotcom** [comment]
As far as swapping out digest names in html files I have been thinking about this and am trying to determine a reliable way to do this. I.e. if its a relative url do we try to resolve like we do with css and resolve in the exact same manner? or do we support a custom url scheme pattern for build time processing i.e. `asset://image.png`

**Comment by rgoldberg** [comment]
Sorry, when I said reroute, I meant that the Grails server would choose the correct asset from the file system (or its own internal cache of files from the file system), not that it would send some sort of HTTP 3xx response to a client.

**Comment by davydotcom** [comment]
not sure if you were aware of this but when you use `grails.assets.storagePath=` on boot we go ahead and copy out both digested and non digested name versions of the files

**Comment by rgoldberg** [comment]
I need to get Asset Pipeline running quickly with static files linking to other static files.

Doing it sans cache digest fingerprints seemed like the quickest solution, but I'm open to anything that will work.

It sounds like the storagePath solution is the fastest way to achieve what I want.

Would it ever be a problem if GSPs using Asset Pipeline refer to the cache-digested versions, while the static files refer to the non-cache-digested versions?  (besides downloading the same file twice, since it's served under 2 different URLs).  Obviously, non-cache-digested assets have versioning issues, but those issues would also exist if Asset Pipeline supported non-cache-digested file names.

If I got the omission of cache digests from asset file names running throughout Asset Pipeline, would there be any issues with it besides the potential versioning issues (which are also present in the storagePath solution)?  This would save any issues from serving the same file from two different URLs.

Thanks for all the help.

**Comment by davydotcom** [comment]
It would work fine, the real issue though is you really need html to reference digested names and all your problems would be solved

**Comment by rgoldberg** [comment]
Yeah, I totally agree that the longterm solution is to have HTML, & JavaScript, also rewrite URLs, but I assume that will take too long to design & code for a quick turnaround.

I'm happy to discuss various options for URL rewriting, as we both seem to have a few ideas about possible solutions, but I'm gonna first concentrate on the storagePath (or some similar quick option).

How long do you think it would take to release a HTML & JS (& possibly more) URL-rewriting version of Asset Pipeline?  You're probably busy with other work besides Asset Pipeline, so I don't expect it anytime soon (but if you had free time, soon would be great).

**Comment by rgoldberg** [comment]
Something like the following project might be very useful:

https://github.com/tednaleid/groovy-template-grails-asset-pipeline

They already support JS & CSS, so just adding HTML support might take care of most people's needs.

**Comment by davydotcom** [comment]
I already have a working version for html relative path replacement working swimmingly

**Comment by rgoldberg** [comment]
Is that committed to a branch so I can try it, and possibly mimic it for JavaScript?

**Comment by davydotcom** [comment]
I havent done this for js yet but you can try it out it is published on version `2.3.0`
